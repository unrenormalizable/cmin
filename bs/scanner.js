export const TOK_ERROR = 0
export const TOK_EOF = 1
export const TOK_WHITESPACE = 2
export const TOK_KEYWORD = 3
export const TOK_IDENTIFIER = 4
export const TOK_SYMBOL = 4

function getCurrChar(text, state) {
  return state.pos + 0 < text.length ? text[state.pos + 0] : null
}

function getNextChar(text, state) {
  return state.pos + 1 < text.length ? text[state.pos + 1] : null
}

function isWhitespace(char) {
  return char !== null && char.match(/\s/)
}

function isAlphabetic(char) {
  return char !== null && char.match(/[A-Za-z_]/)
}

function isNumeric(char) {
  return char !== null && char.match(/[0-9]/)
}

function isAlphanumeric(char) {
  return isAlphabetic(char) || isNumeric(char)
}

function isSymbol(char) {
  return char !== null && char.match(/[\(\)\{\}]/)
}

function isKeyword(str) {
  return (str === 'fn')
}

export function scan(text, state) {
  state = { ...state }

  // WHITESPACE token.
  while (isWhitespace(getCurrChar(text, state))) {
    while (true) {
      const currChar = getCurrChar(text, state)
      const nextChar = getNextChar(text, state)

      if (currChar === '\t' || currChar === ' ') {
        state.pos++
        state.column++
        continue
      }

      if ((currChar === '\r' && nextChar === '\n')) {
        state.pos += 2
        state.line++
        state.column = 1
        continue
      }

      if (currChar === '\r' || currChar === '\n') {
        state.pos++
        state.line++
        state.column = 1
        continue
      }

      break
    }
  }

  // EOF token.
  if (getCurrChar(text, state) === null) {
    return { token: { id: TOK_EOF, value: null, location: state }, state }
  }

  // KEYWORD or IDENTIFIER token.
  if (isAlphabetic(getCurrChar(text, state))) {
    const location = { ...state }
    while (isAlphanumeric(getCurrChar(text, state))) {
      state.pos++
      state.column++
    }

    const value = text.slice(location.pos, state.pos)
    const id = isKeyword(value) ? TOK_KEYWORD : TOK_IDENTIFIER
    return { token: { id, value, location }, state }
  }

  // SYMBOL token.
  if (isSymbol(getCurrChar(text, state))) {
    const location = { ...state }
    state.pos++
    state.column++
    return { token: { id: TOK_SYMBOL, value: text.slice(location.pos, state.pos), location }, state }
  }

  // ERROR token.
  return { token: { id: TOK_ERROR, value: null, location: state }, state }
}

export function scanAll(text) {
  let ret = { state: { pos: 0, line: 1, column: 1 } }
  const tokens = []
  while (true) {
    ret = scan(text, ret.state)
    tokens.push(ret.token)
    if (ret.token.id === TOK_EOF || ret.token.id === TOK_ERROR) {
      break
    }
  }

  return tokens
}
