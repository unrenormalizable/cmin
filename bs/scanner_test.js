import { assertSnapshot } from '@std/testing/snapshot'
import * as scanner from './scanner.js'

Deno.test({
  name: 'scan: empty file',
  fn: async (tc) => {
    const text = ''
    const tokens = scanner.scanAll(text)

    await assertSnapshot(tc, tokens, {})
  },
})

Deno.test({
  name: 'scan: random error token',
  fn: async (tc) => {
    const text = '   \r#  \r\n  \n  '
    const tokens = scanner.scanAll(text)

    await assertSnapshot(tc, tokens, {})
  },
})

Deno.test({
  name: 'scan: empty file with whitespace',
  fn: async (tc) => {
    const text = '   \r\t  \r\n  \n  '
    const tokens = scanner.scanAll(text)

    await assertSnapshot(tc, tokens, {})
  },
})

Deno.test({
  name: 'scan: trivial main function',
  fn: async (tc) => {
    const text = 'fn main() {\n}'
    const tokens = scanner.scanAll(text)

    await assertSnapshot(tc, tokens, {})
  },
})

Deno.test({
  name: 'incremental scanner',
  fn: async (tc) => {
    const text = 'fn main() {\n}'

    let ret = { state: { pos: 0, line: 1, column: 1 } }
    const tokens = []
    while (true) {
      ret = scanner.scan(text, ret.state)
      tokens.push(ret.token)
      if (ret.token.id === scanner.TOK_EOF || ret.token.id === scanner.TOK_ERROR) {
        break
      }
    }

    await assertSnapshot(tc, tokens, {})
  },
})
