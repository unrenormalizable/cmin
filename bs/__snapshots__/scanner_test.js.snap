export const snapshot = {};

snapshot[`scan: empty file 1`] = `
[
  {
    id: 1,
    location: {
      column: 1,
      line: 1,
      pos: 0,
    },
    value: null,
  },
]
`;

snapshot[`scan: random error token 1`] = `
[
  {
    id: 0,
    location: {
      column: 1,
      line: 2,
      pos: 4,
    },
    value: null,
  },
]
`;

snapshot[`scan: empty file with whitespace 1`] = `
[
  {
    id: 1,
    location: {
      column: 3,
      line: 4,
      pos: 14,
    },
    value: null,
  },
]
`;

snapshot[`scan: trivial main function 1`] = `
[
  {
    id: 3,
    location: {
      column: 1,
      line: 1,
      pos: 0,
    },
    value: "fn",
  },
  {
    id: 4,
    location: {
      column: 4,
      line: 1,
      pos: 3,
    },
    value: "main",
  },
  {
    id: 4,
    location: {
      column: 8,
      line: 1,
      pos: 7,
    },
    value: "(",
  },
  {
    id: 4,
    location: {
      column: 9,
      line: 1,
      pos: 8,
    },
    value: ")",
  },
  {
    id: 4,
    location: {
      column: 11,
      line: 1,
      pos: 10,
    },
    value: "{",
  },
  {
    id: 4,
    location: {
      column: 1,
      line: 2,
      pos: 12,
    },
    value: "}",
  },
  {
    id: 1,
    location: {
      column: 2,
      line: 2,
      pos: 13,
    },
    value: null,
  },
]
`;

snapshot[`incremental scanner 1`] = `
[
  {
    id: 3,
    location: {
      column: 1,
      line: 1,
      pos: 0,
    },
    value: "fn",
  },
  {
    id: 4,
    location: {
      column: 4,
      line: 1,
      pos: 3,
    },
    value: "main",
  },
  {
    id: 4,
    location: {
      column: 8,
      line: 1,
      pos: 7,
    },
    value: "(",
  },
  {
    id: 4,
    location: {
      column: 9,
      line: 1,
      pos: 8,
    },
    value: ")",
  },
  {
    id: 4,
    location: {
      column: 11,
      line: 1,
      pos: 10,
    },
    value: "{",
  },
  {
    id: 4,
    location: {
      column: 1,
      line: 2,
      pos: 12,
    },
    value: "}",
  },
  {
    id: 1,
    location: {
      column: 2,
      line: 2,
      pos: 13,
    },
    value: null,
  },
]
`;
